File: API/1040_Mean_Deviation_Index/CS/MeanDeviationIndexStrategy.cs
using System;
using System.Linq;
using System.Collections.Generic;

using Ecng.Common;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

public class MeanDeviationIndexStrategy : Strategy
{
	private readonly StrategyParam<int> _emaPeriod;
	private readonly StrategyParam<int> _atrPeriod;
	private readonly StrategyParam<decimal> _atrMultiplier;
	private readonly StrategyParam<decimal> _level;
	private readonly StrategyParam<int> _signalCooldownBars;
	private readonly StrategyParam<DataType> _candleType;

	private ExponentialMovingAverage _ema;
	private AverageTrueRange _atr;
	private decimal _previousMdx;
	private bool _hasPreviousMdx;
	private int _barsFromSignal;

	public int EmaPeriod { get => _emaPeriod.Value; set => _emaPeriod.Value = value; }
	public int AtrPeriod { get => _atrPeriod.Value; set => _atrPeriod.Value = value; }
	public decimal AtrMultiplier { get => _atrMultiplier.Value; set => _atrMultiplier.Value = value; }
	public decimal Level { get => _level.Value; set => _level.Value = value; }
	public int SignalCooldownBars { get => _signalCooldownBars.Value; set => _signalCooldownBars.Value = value; }
	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }

	public MeanDeviationIndexStrategy()
	{
		_emaPeriod = Param(nameof(EmaPeriod), 26)
			.SetGreaterThanZero()
			.SetDisplay("EMA Period", "EMA length", "General");
		_atrPeriod = Param(nameof(AtrPeriod), 26)
			.SetGreaterThanZero()
			.SetDisplay("ATR Period", "ATR length", "General");
		_atrMultiplier = Param(nameof(AtrMultiplier), 1m)
			.SetGreaterThanZero()
			.SetDisplay("ATR Multiplier", "ATR scaling", "General");
		_level = Param(nameof(Level), 1.2m)
			.SetGreaterThanZero()
			.SetDisplay("Level", "Normalized MDI threshold", "General");
		_signalCooldownBars = Param(nameof(SignalCooldownBars), 10)
			.SetGreaterThanZero()
			.SetDisplay("Signal Cooldown Bars", "Minimum bars between entries", "General");
		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(15).TimeFrame())
			.SetDisplay("Candle Type", "Candles timeframe", "General");
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();
		_ema = null;
		_atr = null;
		_previousMdx = 0m;
		_hasPreviousMdx = false;
		_barsFromSignal = 0;
	}

	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);
		StartProtection(null, null);

		_ema = new ExponentialMovingAverage { Length = EmaPeriod };
		_atr = new AverageTrueRange { Length = AtrPeriod };
		_previousMdx = 0m;
		_hasPreviousMdx = false;
		_barsFromSignal = SignalCooldownBars;

		var subscription = SubscribeCandles(CandleType);
		subscription
			.Bind(_ema, _atr, ProcessCandle)
			.Start();
	}

	private void ProcessCandle(ICandleMessage candle, decimal emaValue, decimal atrValue)
	{
		if (candle.State != CandleStates.Finished)
			return;

		if (!IsFormedAndOnlineAndAllowTrading())
			return;

		if (!_ema.IsFormed || !_atr.IsFormed)
			return;

		var atrVal = atrValue * AtrMultiplier;
		if (atrVal <= 0m)
			return;

		var dev = candle.ClosePrice - emaValue;
		var mdx = dev / atrVal;

		var crossedUp = _hasPreviousMdx && _previousMdx <= Level && mdx > Level;
		var crossedDown = _hasPreviousMdx && _previousMdx >= -Level && mdx < -Level;
		_previousMdx = mdx;
		_hasPreviousMdx = true;

		_barsFromSignal++;
		if (_barsFromSignal < SignalCooldownBars)
			return;

		if (crossedUp && Position <= 0)
		{
			BuyMarket();
			_barsFromSignal = 0;
		}
		else if (crossedDown && Position >= 0)
		{
			SellMarket();
			_barsFromSignal = 0;
		}
	}
}

File: API/1001-1100/1040_Mean_Deviation_Index/PY/mean_deviation_index_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import ExponentialMovingAverage, AverageTrueRange
from StockSharp.Algo.Strategies import Strategy


class mean_deviation_index_strategy(Strategy):
    def __init__(self):
        super(mean_deviation_index_strategy, self).__init__()
        self._ema_period = self.Param("EmaPeriod", 26) \
            .SetGreaterThanZero() \
            .SetDisplay("EMA Period", "EMA length", "General")
        self._atr_period = self.Param("AtrPeriod", 26) \
            .SetGreaterThanZero() \
            .SetDisplay("ATR Period", "ATR length", "General")
        self._atr_multiplier = self.Param("AtrMultiplier", 1.0) \
            .SetGreaterThanZero() \
            .SetDisplay("ATR Multiplier", "ATR scaling", "General")
        self._level = self.Param("Level", 1.2) \
            .SetGreaterThanZero() \
            .SetDisplay("Level", "Normalized MDI threshold", "General")
        self._signal_cooldown_bars = self.Param("SignalCooldownBars", 10) \
            .SetGreaterThanZero() \
            .SetDisplay("Signal Cooldown Bars", "Minimum bars between entries", "General")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(15))) \
            .SetDisplay("Candle Type", "Candles timeframe", "General")
        self._previous_mdx = 0.0
        self._has_previous_mdx = False
        self._bars_from_signal = 0

    @property
    def candle_type(self):
        return self._candle_type.Value

    @candle_type.setter
    def candle_type(self, value):
        self._candle_type.Value = value

    def OnReseted(self):
        super(mean_deviation_index_strategy, self).OnReseted()
        self._previous_mdx = 0.0
        self._has_previous_mdx = False
        self._bars_from_signal = 0

    def OnStarted2(self, time):
        super(mean_deviation_index_strategy, self).OnStarted2(time)
        self._previous_mdx = 0.0
        self._has_previous_mdx = False
        self._bars_from_signal = self._signal_cooldown_bars.Value
        self._ema = ExponentialMovingAverage()
        self._ema.Length = self._ema_period.Value
        self._atr = AverageTrueRange()
        self._atr.Length = self._atr_period.Value
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(self._ema, self._atr, self.OnProcess).Start()

    def OnProcess(self, candle, ema_value, atr_value):
        if candle.State != CandleStates.Finished:
            return
        if not self._ema.IsFormed or not self._atr.IsFormed:
            return
        ev = float(ema_value)
        av = float(atr_value) * float(self._atr_multiplier.Value)
        if av <= 0.0:
            return
        close = float(candle.ClosePrice)
        dev = close - ev
        mdx = dev / av
        lv = float(self._level.Value)
        crossed_up = self._has_previous_mdx and self._previous_mdx <= lv and mdx > lv
        crossed_down = self._has_previous_mdx and self._previous_mdx >= -lv and mdx < -lv
        self._previous_mdx = mdx
        self._has_previous_mdx = True
        self._bars_from_signal += 1
        cd = self._signal_cooldown_bars.Value
        if self._bars_from_signal < cd:
            return
        if crossed_up and self.Position <= 0:
            self.BuyMarket()
            self._bars_from_signal = 0
        elif crossed_down and self.Position >= 0:
            self.SellMarket()
            self._bars_from_signal = 0

    def CreateClone(self):
        return mean_deviation_index_strategy()
